# Contents

1. Scrape and tokenize democratic statements
2. Scrape and tokenize republican statements
3. Build a model and test
4. Make up some phrases.

5. I wanted to semantically map the embedding layer, but due to a bug in this code I have to do it another way. 

6. - Download the "stances" of different campaign sites, and see if this provides a grade of how dem or rep that person is. 

In [1]:
#import dependencies 
from urllib.request import urlopen
from bs4 import BeautifulSoup
import re
import codecs
import nltk
import numpy as np
import pandas as pd
%pylab inline
nltk.download("punkt")
nltk.download("stopwords")

Populating the interactive namespace from numpy and matplotlib
[nltk_data] Downloading package punkt to
[nltk_data]     /Users/noahkasmanoff/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/noahkasmanoff/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

# 1. Scraping various sources for democratic and republican statements/stances. 

In [2]:
#import democratic data

#scrape DNC website
url = "https://democrats.org/about/party-platform/"
client = urlopen(url)
page_html = client.read()
page_soup = BeautifulSoup(page_html,'html.parser')


dem_corpus = []
print("len of dem corpus: ", len(dem_corpus))
for tag in page_soup.find_all(['h1', 'p']):
  #  print(len(tag.text))
  #  print(tag.text)
  #  print()
    dem_corpus.append(tag.text)
print("len of dem corpus: ", len(dem_corpus))

#scrape this website a lot!
dem_urls = ["https://www.republicanviews.org/democratic-views-on-military-spending/",
       "https://www.republicanviews.org/democratic-views-on-gay-marriage/",
       "https://www.republicanviews.org/democratic-party-beliefs/",
       "https://www.republicanviews.org/democratic-views-on-planned-parenthood/",
       "https://www.republicanviews.org/democratic-views-on-foreign-policy/",
       "https://www.republicanviews.org/democratic-views-on-illegal-immigration/",
       "https://www.republicanviews.org/democratic-views-on-banking-regulation/",
       "https://www.republicanviews.org/democratic-views-on-crime/",
       "https://www.republicanviews.org/democratic-views-on-the-u-s-embassy-in-jerusalem/",
       "https://www.republicanviews.org/democratic-views-on-trade/",
       "https://www.republicanviews.org/democratic-views-on-energy/",
       "https://www.republicanviews.org/democratic-views-on-national-security/",
       "https://www.republicanviews.org/democratic-views-on-the-military/"]


for url in dem_urls:
    client = urlopen(url)
    page_html = client.read()
    page_soup = BeautifulSoup(page_html,'html.parser')
    tag = page_soup.findAll('article')[0]  #that's it. 
    dem_corpus.append(tag.text)
    print("len of dem corpus: ", len(dem_corpus))

#scrape some politics website. Does not use other indexing! 
url = "http://www.ontheissues.org/Democratic_Party.htm"
client = urlopen(url)
page_html = client.read()
page_soup = BeautifulSoup(page_html,'html.parser')


for tag in page_soup.find_all(['li']):
    dem_corpus.append(tag.text)
print("len of dem corpus: ", len(dem_corpus))



#via wikipedia
url = "https://en.wikipedia.org/wiki/Political_positions_of_the_Democratic_Party"

client = urlopen(url)
page_html = client.read()
page_soup = BeautifulSoup(page_html,'html.parser')


for tag in page_soup.find_all(['h1', 'p']):
    dem_corpus.append(tag.text)
print("len of dem corpus: ", len(dem_corpus))



#scrape senate democrats website
url = "https://www.democrats.senate.gov/abetterdeal/higher-wages-and-better-jobs"

client = urlopen(url)
page_html = client.read()
page_soup = BeautifulSoup(page_html,'html.parser')


for tag in page_soup.find_all(['h1', 'p']):
  #  print(len(tag.text))
  #  print(tag.text)
  #  print()
    dem_corpus.append(tag.text)
print("len of dem corpus: ", len(dem_corpus))



#continue scraping senate democrats
url = "https://www.democrats.senate.gov/abetterdeal/lower-the-cost-of-living-for-families"

client = urlopen(url)
page_html = client.read()
page_soup = BeautifulSoup(page_html,'html.parser')

for tag in page_soup.find_all(['h1', 'p']):

    dem_corpus.append(tag.text)
print("len of dem corpus: ", len(dem_corpus))


#and again 

url = "https://www.democrats.senate.gov/abetterdeal/tools-to-succeed-in-the-21st-century"

client = urlopen(url)
page_html = client.read()
page_soup = BeautifulSoup(page_html,'html.parser')

for tag in page_soup.find_all(['h1', 'p']):

    dem_corpus.append(tag.text)
print("len of dem corpus: ", len(dem_corpus))

#combine all elements into a continuous body, now resembling a corpus. 
dem_corpus = " ".join(dem_corpus)


tokenizer = nltk.data.load('tokenizers/punkt/english.pickle')


raw_dem_sentences = tokenizer.tokenize(dem_corpus)

def sentence_to_wordlist(raw):
    clean = re.sub("[^a-zA-Z]"," ", raw)
    words = clean.split()
    return words


dem_sentences = []
for raw_dem_sentence in raw_dem_sentences:
    if len(raw_dem_sentence) > 0:
        dem_sentences.append(sentence_to_wordlist(raw_dem_sentence))
print()     
print("Here's an example datapoint.")      
print(raw_dem_sentences[50])
print(sentence_to_wordlist(raw_dem_sentences[50]))



token_count = sum([len(dem_sentence) for dem_sentence in dem_sentences])
print("The dem policy corpus contains {0:,} tokens".format(token_count))


y_dems = np.ones(shape=np.shape(dem_sentences)) #now the labels. 
print("Done!")

len of dem corpus:  0
len of dem corpus:  282
len of dem corpus:  283
len of dem corpus:  284
len of dem corpus:  285
len of dem corpus:  286
len of dem corpus:  287
len of dem corpus:  288
len of dem corpus:  289
len of dem corpus:  290
len of dem corpus:  291
len of dem corpus:  292
len of dem corpus:  293
len of dem corpus:  294
len of dem corpus:  295
len of dem corpus:  460
len of dem corpus:  502
len of dem corpus:  517
len of dem corpus:  528
len of dem corpus:  539

Here's an example datapoint.
We believe in protecting civil liberties and guaranteeing civil rights and voting rights, women’s rights and workers’ rights, LGBT rights, and rights for people with disabilities.
['We', 'believe', 'in', 'protecting', 'civil', 'liberties', 'and', 'guaranteeing', 'civil', 'rights', 'and', 'voting', 'rights', 'women', 's', 'rights', 'and', 'workers', 'rights', 'LGBT', 'rights', 'and', 'rights', 'for', 'people', 'with', 'disabilities']
The dem policy corpus contains 55,885 tokens
Done!


In [3]:
#Now the republican data. 

#scrape wikipedia description. 
url = "https://en.wikipedia.org/wiki/Political_positions_of_the_Republican_Party"
client = urlopen(url)
page_html = client.read()
page_soup = BeautifulSoup(page_html,'html.parser')


rep_corpus = []
print("len of rep corpus: ", len(rep_corpus))
for tag in page_soup.find_all(['h1', 'p']):
  #  print(len(tag.text))
  #  print(tag.text)
  #  print()
    rep_corpus.append(tag.text)
print("len of rep corpus: ", len(rep_corpus))

#Scrape GOP website for official platform
#Forbidden, so I copy pasted it all into a text file also in this repo. 

#on the issues website
url = "http://www.ontheissues.org/Republican_Party.htm"

client = urlopen(url)
page_html = client.read()
page_soup = BeautifulSoup(page_html,'html.parser')


for tag in page_soup.find_all(['li']):
  #  print(len(tag.text))
  #  print(tag.text)
  #  print()
    rep_corpus.append(tag.text)
print("len of rep corpus: ", len(rep_corpus))



rep_urls  = ["https://www.republicanviews.org/republican-views-on-the-economy/",

"https://www.republicanviews.org/republican-views-on-energy/",

"https://www.republicanviews.org/republican-views-on-the-environment/",

"https://www.republicanviews.org/republican-views-on-health-care/",

"https://www.republicanviews.org/republican-views-on-immigration/",

"https://www.republicanviews.org/republican-views-on-social-security/",

"https://www.republicanviews.org/republican-views-on-bitcoin-and-cryptocurrencies/",

"https://www.republicanviews.org/republican-views-on-the-u-s-embassy-in-jerusalem/",

"https://www.republicanviews.org/republican-views-on-unemployment/",

"https://www.republicanviews.org/republican-views-on-terrorism/",

"https://www.republicanviews.org/republican-views-on-stem-cell-research/",

"https://www.republicanviews.org/republican-views-on-the-electoral-college/",

"https://www.republicanviews.org/republican-views-on-religion/",

"https://www.republicanviews.org/republican-views-on-net-neutrality/"]

for url in rep_urls:
    client = urlopen(url)
    page_html = client.read()
    page_soup = BeautifulSoup(page_html,'html.parser')
    tag = page_soup.findAll('article')[0]  #that's it. 
    rep_corpus.append(tag.text)
    print("len of rep corpus: ", len(rep_corpus))



rep_corpus = " ".join(rep_corpus)
tokenizer = nltk.data.load('tokenizers/punkt/english.pickle')




#open the text file and 
rep_platform_corpus_raw = u""
print("Reading '{0}'...".format('rep_platform.txt'))
with codecs.open('rep_platform.txt', "r", "utf-8") as book_file:
    rep_platform_corpus_raw += book_file.read()

rep_corpus += rep_platform_corpus_raw





raw_rep_sentences = tokenizer.tokenize(rep_corpus)


def sentence_to_wordlist(raw):
    clean = re.sub("[^a-zA-Z]"," ", raw)
    words = clean.split()
    return words


rep_sentences = []
for raw_rep_sentence in raw_rep_sentences:
    if len(raw_rep_sentence) > 0:
        rep_sentences.append(sentence_to_wordlist(raw_rep_sentence))
        
print()
print("Here's an example sentence in rep corpus.")
print(raw_rep_sentences[10])
print(sentence_to_wordlist(raw_rep_sentences[10]))


token_count = sum([len(rep_sentence) for rep_sentence in rep_sentences])
print("The rep policy corpus contains {0:,} tokens".format(token_count))


y_reps = np.zeros(shape=np.shape(rep_sentences))


len of rep corpus:  0
len of rep corpus:  72
len of rep corpus:  280
len of rep corpus:  281
len of rep corpus:  282
len of rep corpus:  283
len of rep corpus:  284
len of rep corpus:  285
len of rep corpus:  286
len of rep corpus:  287
len of rep corpus:  288
len of rep corpus:  289
len of rep corpus:  290
len of rep corpus:  291
len of rep corpus:  292
len of rep corpus:  293
len of rep corpus:  294
Reading 'rep_platform.txt'...

Here's an example sentence in rep corpus.
[14]
 Republican party leaders strongly believe that free markets and individual achievement are the primary factors behind economic prosperity.
['Republican', 'party', 'leaders', 'strongly', 'believe', 'that', 'free', 'markets', 'and', 'individual', 'achievement', 'are', 'the', 'primary', 'factors', 'behind', 'economic', 'prosperity']
The rep policy corpus contains 78,822 tokens


# Now it's modelling time 

Reference: 
https://github.com/Hvass-Labs/TensorFlow-Tutorials/blob/master/20_Natural_Language_Processing.ipynb

In [4]:
%matplotlib inline
import matplotlib.pyplot as plt
import tensorflow as tf
import numpy as np
from scipy.spatial.distance import cdist


# from tf.keras.models import Sequential  # This does not work!
from tensorflow.python.keras.models import Sequential
from tensorflow.python.keras.layers import Dense, GRU, Embedding
from tensorflow.python.keras.optimizers import Adam
from tensorflow.python.keras.preprocessing.text import Tokenizer
from tensorflow.python.keras.preprocessing.sequence import pad_sequences

# 2. Tokenizing and splitting these statements into a training/testing format. 

In [5]:
raw_rep_sentence

'A Republican administration will champion an open and free internet based on principles of free expression and universal values and will pursue policies to empower citizens and U.S. companies operating in authoritarian countries to circumvent internet firewalls and gain accurate news and information online.'

To recreate the structure of the imdb dataset, the next plan is to concatenate the dem and rep texts and simultaneously with the labels of 1 or 0, and probably shuffle both. 

In [6]:
X_text = np.concatenate((raw_dem_sentences,raw_rep_sentences))
y = np.concatenate((y_dems,y_reps))

In [7]:
from sklearn.model_selection import train_test_split

X_train_text, X_test_text, y_train, y_test = train_test_split(X_text, y ,shuffle=True, test_size=0.33, random_state=42)

In [8]:
len(X_text)

6659

In [9]:
token_count = sum([len(sentence) for sentence in X_text])
print("The rep policy corpus contains {0:,} tokens".format(token_count))


The rep policy corpus contains 886,886 tokens


The training and testing data are now separated and shuffled. Next task is to tokenize these words once again, but now this will be based on an integer value for indexing each word (note this contains no semantic meaning), and once fit you can determine how many unique words to include.

In [10]:
num_words=8000
tokenizer = Tokenizer(num_words=num_words)


Fitting: 
The tokenizer scans through all the text and strips unwanted characters, converts everything to lowercase, and builds a vocabulary of all the unique words. It's ok that it uses the entire dataset, as the embedding is only based on the training data (not sure how accurate this is though should follow this up). 

In [11]:
tokenizer.fit_on_texts(X_text)


In [12]:
if num_words is None:
    num_words = len(tokenizer.word_index)
print(len(tokenizer.word_index))

9597


In [13]:
num_words

8000

In [14]:
#this is ordered by the number of occurences of the word. 
tokenizer.word_index


{'the': 1,
 'and': 2,
 'to': 3,
 'of': 4,
 '\r': 5,
 'in': 6,
 'a': 7,
 'that': 8,
 'for': 9,
 'we': 10,
 'on': 11,
 'is': 12,
 'aug': 13,
 'our': 14,
 'as': 15,
 '2000': 16,
 'will': 17,
 'with': 18,
 '2004': 19,
 'by': 20,
 'their': 21,
 'be': 22,
 'are': 23,
 'it': 24,
 'not': 25,
 'have': 26,
 'sep': 27,
 'support': 28,
 'democrats': 29,
 'they': 30,
 'from': 31,
 'republican': 32,
 'more': 33,
 'has': 34,
 'should': 35,
 'or': 36,
 'this': 37,
 'democratic': 38,
 'american': 39,
 'party': 40,
 'republicans': 41,
 'all': 42,
 'views': 43,
 'believe': 44,
 'government': 45,
 'health': 46,
 'federal': 47,
 'an': 48,
 'who': 49,
 'must': 50,
 'military': 51,
 'states': 52,
 'also': 53,
 'rights': 54,
 'national': 55,
 'people': 56,
 '”': 57,
 'at': 58,
 'care': 59,
 'jul': 60,
 '2012': 61,
 'which': 62,
 'security': 63,
 'public': 64,
 'new': 65,
 'americans': 66,
 'those': 67,
 'against': 68,
 'would': 69,
 'its': 70,
 'state': 71,
 'no': 72,
 'other': 73,
 'president': 74,
 'but': 7

In [15]:
#now it's time to tokenize all of the training and testing data for future use. 
X_train_tokens = tokenizer.texts_to_sequences(X_train_text)

X_test_tokens = tokenizer.texts_to_sequences(X_test_text)

# Padding and truncating

Not all senetences are created equal (both in meaning and length!) To preserve the same size input property of a neural network, this means padding the shorter phrases, and truncating the longer ones. For padding these zeros (which will be interpreted as meaningless) will be placed at the start of the sentence as the sequential nature should play less of a role. Rather than an abrupt jump to zero which could have an adverse effect on the RNN, it is smarter to go from 0 up. 

In [16]:
num_tokens = [len(tokens) for tokens in X_train_tokens + X_test_tokens]
num_tokens = np.array(num_tokens)

In [17]:
np.mean(num_tokens)
#avg num of tokens in a phrase

20.730740351404116

In [18]:
np.max(num_tokens)
#max num of tokens in a phrase

213

In [19]:
max_tokens = np.mean(num_tokens) + 2 * np.std(num_tokens)
max_tokens = int(max_tokens)
max_tokens

50

In [20]:
np.sum(num_tokens < max_tokens) / len(num_tokens)


0.96200630725334135

So by setting it to mean + 2 stds, this covers about 93% of the dataset. Fine by me!

In [21]:
pad = 'pre' #pad in the beginning, reasoning explained earlier. 

X_train_pad = pad_sequences(X_train_tokens, maxlen=max_tokens,
                            padding=pad, truncating=pad)

X_test_pad = pad_sequences(X_test_tokens, maxlen=max_tokens,
                           padding=pad, truncating=pad)

In [22]:
X_train_pad.shape

(4461, 50)

In [23]:
X_test_pad.shape


(2198, 50)

In [24]:
idx = tokenizer.word_index
inverse_map = dict(zip(idx.values(), idx.keys()))

In [25]:
def tokens_to_string(tokens):
    """Helper function to go back to the text."""
    # Map from tokens back to words.
    words = [inverse_map[token] for token in tokens if token != 0]
    
    # Concatenate all words.
    text = " ".join(words)

    return text

In [26]:
len(tokenizer.word_index)

9597

In [27]:
X_train_text[10]

'(Sep 2004)\r\n    Provide new strategies to help poor nations.'

In [28]:
tokens_to_string(X_train_tokens[10])


'sep 2004 \r provide new strategies to help poor nations'

It works!

# 3. Creating a Recurrent Neural Network 


Now for the implementation of an RNN to classify this data, using the keras structure. 

In [29]:
model = Sequential()
embedding_size = 10  #dimensions of word2vec (basically.)

#this is essentially a word2vec model but it is a part of the RNN, as it trains simultaneously. 
model.add(Embedding(input_dim=num_words,
                    output_dim=embedding_size,
                    input_length=max_tokens,  
                    name='layer_embedding'))



model.add(GRU(units=16, return_sequences=True))
model.add(GRU(units=8, return_sequences=True))
model.add(GRU(units=4))
model.add(Dense(1, activation='sigmoid'))
optimizer = Adam(lr=1e-3)
model.compile(loss='binary_crossentropy',
              optimizer=optimizer,
              metrics=['accuracy'])

In [30]:
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
layer_embedding (Embedding)  (None, 50, 10)            80000     
_________________________________________________________________
gru_1 (GRU)                  (None, 50, 16)            1296      
_________________________________________________________________
gru_2 (GRU)                  (None, 50, 8)             600       
_________________________________________________________________
gru_3 (GRU)                  (None, 4)                 156       
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 5         
Total params: 82,057
Trainable params: 82,057
Non-trainable params: 0
_________________________________________________________________


In [31]:
model.fit(X_train_pad, y_train,
          validation_split=0.05, epochs=15, batch_size=64)


Train on 4237 samples, validate on 224 samples
Epoch 1/15
4237/4237 [==============================] - 8s 2ms/step - loss: 0.6767 - acc: 0.5754 - val_loss: 0.6481 - val_acc: 0.6473
Epoch 2/15
4237/4237 [==============================] - 6s 1ms/step - loss: 0.6643 - acc: 0.6141 - val_loss: 0.6447 - val_acc: 0.6473
Epoch 3/15
4237/4237 [==============================] - 6s 1ms/step - loss: 0.6518 - acc: 0.6169 - val_loss: 0.6238 - val_acc: 0.6652
Epoch 4/15
4237/4237 [==============================] - 6s 1ms/step - loss: 0.5443 - acc: 0.7274 - val_loss: 0.5331 - val_acc: 0.7545
Epoch 5/15
4237/4237 [==============================] - 6s 1ms/step - loss: 0.4342 - acc: 0.8058 - val_loss: 0.4795 - val_acc: 0.7991
Epoch 6/15
4237/4237 [==============================] - 6s 1ms/step - loss: 0.3375 - acc: 0.8671 - val_loss: 0.4660 - val_acc: 0.8036
Epoch 7/15
4237/4237 [==============================] - 6s 1ms/step - loss: 0.2636 - acc: 0.9063 - val_loss: 0.4953 - val_acc: 0.7902
Epoch 8/15
4237

In [32]:
result = model.evaluate(X_test_pad, y_test)
print("Accuracy: {0:.2%}".format(result[1]))


2198/2198 [==============================] - 1s 516us/step
Accuracy: 83.48%


# Example of mis-classified text



In [33]:
#In order to show an example of mis-classified text, 
#we first calculate the predicted sentiment for the first 500 texts in the test-set.

y_pred = model.predict(x=X_test_pad[0:500])
y_pred = y_pred.T[0]


#These predicted numbers fall between 0.0 and 1.0. 
#We use a cutoff / threshold and say that all values above 0.5 
#are taken to be 1.0 and all values below 0.5 are taken to be 0.0. 
#This gives us a predicted "class" of either 0.0 or 1.0.


cls_pred = np.array([1.0 if p>0.5 else 0.0 for p in y_pred])
cls_true = np.array(y_test[0:500])


We can then get indices for all the texts that were incorrectly classified by comparing all the "classes" of these two arrays.



In [34]:
len(y_dems)

2580

In [35]:
len(y_reps)

4079

In [37]:
len(y_reps) / (len(y_dems) + len(y_reps))

0.6125544376032437

In [38]:
incorrect = np.where(cls_pred != cls_true)
incorrect = incorrect[0]

len(incorrect)  #how many out of 500 were wrong?


76

In [39]:
idx = incorrect[-1]


In [40]:
text = X_test_text[idx]
text

'They also hope to prohibit gun sales to those receiving involuntary mental health services on an outpatient basis, if a court deems them dangerous.'

In [41]:
y_pred[idx]


0.11404887

In [42]:
for val in incorrect:
    print(val)
    idx = val
    text = X_test_text[idx]
    print()
    print(text)
    print("Incorrectly classified as ",y_pred[idx])
    print()

4

While House Democrats such as Nancy Pelosi demand a road to citizenship for illegal immigrants, Boehner’s proposed plan discusses no such accommodation.
Incorrectly classified as  0.982943

5

These are public lands, and the public should have access to them for appropriate activities like hunting, fishing, and recreational shooting.
Incorrectly classified as  0.928313

6

Consumer choice is the most powerful factor in healthcare reform.
Incorrectly classified as  0.781549

12

Democrats believe in reforming asylum policies, believing that we should have “equitable asylum policies that treat people the same whether they have fled violence from the Right and Left.” They believe that immigrants are currently facing a lack of due process protections, and wish to see these protections restored, so that immigrants cannot face deportation due to minor offenses, “and are eligible to receive safety net services supported by their tax dollars.”
Democratic Views On The Death Penalty
Democrati

# New Data

Now let's try and classify new statements. Some will be obvious, while others should be consfusing.  

In [43]:
text1 = "Climate change is the most important issue to address."
text2 = "We need a safer form of gun control."
text3 = "Obamacare is bad"
text4 = "Our welfare system is broken"
text5 = "The wealthy need to pay their fair share in taxes."
text6 = "Roe v Wade must be overturned."
text7 = "Make America Great Again"
text8 =  "Israel is our greatest ally."  #inherently ambiguous (I hope!)
text9 = "I love russia."
texts = [text1, text2, text3, text4, text5, text6, text7, text8,text9]

In [44]:
tokens = tokenizer.texts_to_sequences(texts)


In [45]:

tokens_pad = pad_sequences(tokens, maxlen=max_tokens,
                           padding=pad, truncating=pad)
tokens_pad.shape

(9, 50)

In [46]:
model.predict(tokens_pad)


array([[ 0.96718007],
       [ 0.39847985],
       [ 0.02395949],
       [ 0.01504334],
       [ 0.10007751],
       [ 0.01378797],
       [ 0.89935702],
       [ 0.05324052],
       [ 0.97865218]], dtype=float32)

# Now recreating that prediction method with a raw input for a more direct approach.

In [57]:
text = [input()]
tokens = tokenizer.texts_to_sequences(text)
tokens_pad = pad_sequences(tokens, maxlen=max_tokens,
                           padding=pad, truncating=pad)

x = model.predict(tokens_pad)
if x[0][0] < .5:
    print("Republican")
else:
    print("Democrat")


Democrat


In [58]:
if x[0][0] < .5:
    print("Republican")
else:
    print("Democrat")

Democrat


Qualitatively speaking, this model gueses 6 out of 7, not bad! The democratic ones that serve as the first two are the one's misclassified, and maybe this says something about their policy...

# Semantic Exploration


One perk of the embedding layer is that during the training period, it semantically maps the meaning of different words based on their classification, relative place in the sequence, etc. Certain terms will end up being more meaningful ot the message, while others are not (like stop words, 'and', the',etc.)


By retreiving the embedding layer, we can explore these embeddings and have a little peek under the hood as to what this model has learned about politics. 

In [59]:
#get the layer from the model
layer_embedding = model.get_layer('layer_embedding')


In [60]:
#get the weights from that layer. 
weights_embedding = layer_embedding.get_weights()[0]
print(weights_embedding.shape)

(8000, 10)


In [61]:
for i, vec, in enumerate(weights_embedding):
    #convert to word
    tokenizer.word_index
  

Previously we said that the vec would have 8 dimensions, so this is consistent with those findings here. 

In [62]:
#ex what is the 
token_good = tokenizer.word_index['good']
token_good

207

In [63]:
token_great = tokenizer.word_index['great']
token_great

423

In [64]:
weights_embedding[token_good]


array([ 0.82499659,  0.70365268,  0.87161672,  0.42685714,  0.06458494,
        0.61854428,  0.22443061,  0.57985944,  0.1151525 ,  0.68127704], dtype=float32)

In [65]:
weights_embedding[token_great]


array([ 0.5167501 ,  0.44600865,  0.53695512,  0.25615156,  0.2797631 ,
        0.29351661,  0.23447201,  0.71865511,  0.74952543,  0.16358115], dtype=float32)

Since these words are similar, we'd expect their embeddings to be close (speaking from displacement meaning) We see that in some of these dimensions, that is the case. 

# Sorting

Using this helper function, find the distances between every word and sort.



# There is an issue with this function where the index goes beyond what I have originally set. Seems this won't work (yet). 

In [66]:
def print_sorted_words(word, metric='cosine'):
    """
    Print the words in the vocabulary sorted according to their
    embedding-distance to the given word.
    Different metrics can be used, e.g. 'cosine' or 'euclidean'.
    """

    # Get the token (i.e. integer ID) for the given word.
    token = tokenizer.word_index[word]

    # Get the embedding for the given word. Note that the
    # embedding-weight-matrix is indexed by the word-tokens
    # which are integer IDs.
    embedding = weights_embedding[token]

    # Calculate the distance between the embeddings for
    # this word and all other words in the vocabulary.
    distances = cdist(weights_embedding, [embedding],
                      metric=metric).T[0]
    
    # Get an index sorted according to the embedding-distances.
    # These are the tokens (integer IDs) for words in the vocabulary.
    sorted_index = np.argsort(distances)
    
    # Sort the embedding-distances.
    sorted_distances = distances[sorted_index]
    
    # Sort all the words in the vocabulary according to their
    # embedding-distance. This is a bit excessive because we
    # will only print the top and bottom words.
    sorted_words = [inverse_map[token] for token in sorted_index
                  if token != 0]
    # Helper-function for printing words and embedding-distances.
    def _print_words(words, distances):
        for word, distance in zip(words, distances):
            print("{0:.3f} - {1}".format(distance, word))

    # Number of words to print from the top and bottom of the list.
    k = 10

    print("Distance from '{0}':".format(word))

    # Print the words with smallest embedding-distance.
    _print_words(sorted_words[1:k+1], sorted_distances[1:k+1])

    print("...")

    # Print the words with highest embedding-distance.
    _print_words(sorted_words[-k:], sorted_distances[-k:])

In [69]:
print_sorted_words('obamacare', metric='cosine')


Distance from 'obamacare':
0.011 - institutions
0.015 - commander
0.016 - commissions
0.018 - miners
0.018 - statehood
0.020 - burden
0.020 - fcc’s
0.021 - commend
0.021 - months
0.022 - “building
...
0.465 - august
0.466 - narco
0.468 - suspected
0.473 - couples
0.482 - stronger
0.483 - credits
0.483 - threatened
0.519 - unrivaled
0.588 - democrats
0.845 - jul


# New Idea: I've downloaded Josh Gottheimer's stances on the issues, now tokenize and model this to predict each statement, looking at the strongest ones on each as proof.

In [70]:
#open the text file and 
gottheimer_raw = u""
print("Reading '{0}'...".format('gottissues.txt'))
with codecs.open('gottissues.txt', "r", "utf-8") as book_file:
    gottheimer_raw += book_file.read()




Reading 'gottissues.txt'...


In [71]:
gottheimer_raw  #now convert !

'WORKING FOR YOU!\nIn Congress, I’m laser-focused on getting things done for our families and for our children. When I got there, I quickly noticed that the extreme partisanship and gridlock were even worse than any of us could have imagined. However, I also found that from day one, if you actually made an effort to reach across the aisle, to find common ground, you could actually get things done that are good for the country and for New Jersey.\n\nEarly last year, I was elected Co-Chair of the bipartisan Problem Solvers Caucus. It’s a group of 24 Democrats and 24 Republicans who meet every week to see where we can make progress on our toughest issues, from lowering taxes, to creating jobs, to fixing our infrastructure, to balancing the budget. Just like when I was at Ford Motor Company and Microsoft, my goal is to bring people together to get something done. I was honored to be recognized as the most bipartisan freshman Democrat in Congress. For me, it’s not about a partisan agenda; i

In [72]:
gottheimer_text = gottheimer_raw.split('.')

From what I can tell, this data needs to already be in the corpus prior to testing. This means re-running the model every time you grab one of these text files. I need a robust scraper that immediately gets all the text including this one, but keeping them in separate directories until testing later. This is especially because this prediction data is not labelled. 

All the blank spaces correspond to words not in original corpus... 

In [73]:
gottheimer_tokens = tokenizer.texts_to_sequences(gottheimer_text)

In [74]:
gottheimer_tokens_pad = pad_sequences(gottheimer_tokens, maxlen=max_tokens,
                            padding=pad, truncating=pad)

In [75]:
results = model.predict(gottheimer_tokens_pad)

In [76]:
x = 0
for result in results:
    if result > .5:
        x+=1
    

In [78]:
x/ len(results)

0.6

105 of the len(results)= 135 statements are democratic, meaning this is a democratic candidate (probably )

# Now to repeat this testing this with a republican candidate, for reproducibility, I'll stick with Gott's competitor, John Mcann


In [79]:
#open the text file and 
mcann_raw = u""
print("Reading '{0}'...".format('mcissues.txt'))
with codecs.open('mcissues.txt', "r", "utf-8") as book_file:
    mcann_raw += book_file.read()


mcann_text = mcann_raw.split('.')

Reading 'mcissues.txt'...


In [80]:
mcann_tokens = tokenizer.texts_to_sequences(mcann_text)

In [81]:
mcann_tokens_pad = pad_sequences(mcann_tokens, maxlen=max_tokens,
                            padding=pad, truncating=pad)

In [82]:
results = model.predict(mcann_tokens_pad)

In [83]:
x = 0
for result in results:
    if result > .5:
        x+=1
    

In [87]:
print(x/len(results))

0.48598130841121495


In [85]:
len(results)

107

In [86]:
mcann_text

['Too many Americans are unemployed, underemployed or working undesirable jobs',
 '\n\nThat’s why I’m running for Congress',
 '\n\nTaxes\n\nI will work to make the tax code more favorable to small businesses which are the backbone of the nation’s economy',
 ' It’s time that Congress put Main Street before Wall Street',
 '\n\nSince historic tax cuts passed, New Jersey residents are concerned about the elimination of the SALT (state and local tax) deduction',
 ' Elect me to Congress and I will work to fix this pertinent concern',
 '\n\nTrade\n\nWe’ve all heard the arguments for free and fair trade',
 ' Like the President, I am for America First trade',
 ' We need to reevaluate trade deals to see what benefits America first',
 '\n\nWe’ve been taken advantage of too often on the world stage—while your jobs have been shipped off to other countries',
 '\n\nAs your Congressman I will never vote for a bill that guts American jobs and creates a negative trade balance',
 ' All trade deals need t

# Conclusion 


By looking at the issues page of a republican and democrat running for the same seat, we see that of the 135 detected statements on the dem's page, 105 were considered dem, good for 78%. On the contrary, the republican's issue page has 107 detected statements, where 61 register as democrat or 57%. At face value this means that the model successfully distinguished which candidate is more democratic of the two, there is a lot more work to be done. For one the training data isn't necessarily perfect and could be drawn from more sources, the strengh of such statements used here may have an impact, among other things. Overall this seems like a good starting point and this project seems to have some potential. 